### PR TITLE
Fixes and Improvements to the LoRA training.

### PR DESCRIPTION
**Pull request by Nan-Do**
In this pull request I'm sharing some small fixes and improvements I have implemented to be able to train and use LoRAs.
Changes to LoRA.py:
- PeftConfig doesn't take a dtype and device_map parameters so the code fails when loading a LoRA using a model loaded in 8 or 4 bits. The fix checks if the model was loaded in 4 or 8 bits and if that's the case it uses the appropriate type name.
- https://huggingface.co/docs/peft/package_reference/config#peft.PeftConfig
 
Changes to Training.py:
- Adding the parameter skip_special_tokens=True when decoding raw_text in the text preparation phase, without this flag the special tokens will remain in the text during the preparation phase and this will make the training to completely fail. 
    - Sample output using the LlamaTokenizer (after decoding):
 `<s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi L'`
    - Sample output of the training process:
`{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.02}`
`{'train_runtime': 2932.1946, 'train_samples_per_second': 5.752, 'train_steps_per_second': 0.09, 'train_loss': 0.0, 'epoch': 0.02}`
- I have modified the chunking process, this doesn't fail but the first generated chunk will be shorter and a the text it contains will be repeated,  also the change makes the chunking process substantially more efficient both in memory and time. 
    - Sample of the first few generated chunks (after decoding). Before the change:
$1^{st}$:`Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi L`
$2^{nd}$:`Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehender`
$3^{rd}$:`orem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostr`

   - and after:
$1^{st}$:`Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehender`
$2^{nd}$:`orem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostr`

The added file:
 - [vicuna-format.json](https://github.com/oobabooga/text-generation-webui/compare/main...Nan-Do:text-generation-webui:main?expand=1#diff-4b5f665509464c69e1f6fbe4d5fac7e9a5b301efcdc3230e210fe769fa018f06) Is the format to fine-tune Vicuna1.1 models which is the most efficient model at 7B parameters

### CODE REVIEW

**Comment by oobabooga** [comment]
About `skip_special_tokens=True`: could there be situations where the user manually enters tokens like <s> and </s> into the text, for instance to properly include multiple unrelated documents in a single text file, and this addition leads to a loss of those tokens?

**Comment by oobabooga** [comment]
I'd like to invite @FartyPants for review

**Comment by Nan-Do** [comment]
@oobabooga Yeah that makes sense, and I hope I'm not missing something, but the text gets retokenized just before creating the dataset. The tokenize and encode functions already take care of adding the bos token so in the end, and for the LlamaTokenizer, the bos token will be added several times for each prompt. 

Also upon further inspection I have found another potential problem.
The tokenize function will be called for each chunk and depending on the cutoff_len it might happen that most of the chunks don't contain the train_only_after string which will make the chunk be used for training regardless. I'm thinking, for example, of a case in which the instruction or input part of the prompt are longer than cutoff_len, in that case the instruction and/or input part of the prompt will be used for training which might not be the desired outcome of specifying the train_only_after option.

``` python
def encode(text, add_bos_token):
    result = shared.tokenizer.encode(text, truncation=True, max_length=cutoff_len)
    if not add_bos_token and result[0] == shared.tokenizer.bos_token_id:
        result = result[1:]
    return result

def tokenize(prompt):

    if train_only_after == '' or train_only_after not in prompt:
        input_ids = encode(prompt, True)
        input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
        labels = [1] * len(input_ids)

    else:
        ind = prompt.index(train_only_after) + len(train_only_after)
        before_tokens = encode(prompt[:ind], True)
        after_tokens = encode(prompt[ind:], False)

        full_length = len(after_tokens) + len(before_tokens)
        if full_length > cutoff_len:
            after_tokens = after_tokens[:cutoff_len - len(before_tokens)]
        else:
            before_tokens = [shared.tokenizer.pad_token_id] * (cutoff_len - full_length) + before_tokens

        input_ids = before_tokens + after_tokens
        labels = [-100] * len(before_tokens) + [1] * len(after_tokens)

    input_ids = torch.tensor(input_ids)
    return {
        "input_ids": input_ids,
        "labels": labels,
        "attention_mask": input_ids.ne(shared.tokenizer.pad_token_id),
    }
```

**Comment by FartyPants** [comment]
Do we know why: 
without this flag the special tokens will remain in the text during the preparation phase and this will make the training to completely fail. 
(Note: I had seen this behavior when I ended up with training with 'train_loss': 0.0 - but didn't connected it with having token)

As for necessity having <s> </s> in plaintext if it breaks training - then no, obviously. We already have Hard Cut String that should be better to utilize for the purpose of adding arbitrary EOS. - > but I feel I'm in a loop here getting back to the original question - Why and When having special token breaks training? 

As for the train_only_after, yes, correct, but this makes sense for dataset only (I assume to get rid of ### Response and such as written on the UI) and not for plaintext. So if someone is finetuning with a dataset where items cross cutoff_len then that person has a far bigger problem to worry about - he is likely producing garbage with or without train_only_after

**Comment by FartyPants** [comment]
I sprinkled pride and prejudice with random EOS `</s>` and the training definitely didn't care as it obviously shouldn't.

I think the OP problem lies somewhere else.

**Comment by Nan-Do** [comment]
@FartyPants I jumped to conclusions too fast for the reason of failing and my explanation is quite lackluster there. I definitely wasn't thorough enough explaining that part and shouldn't have mentioned that the train fails because of that. 

Let me rephrase it, I assumed that the problem was with the BOS token, right now it will appear several times at the begging of the text and possibly again later as the chunking process will copy it if the flag is on. If I am not mistaken the training text will end looking up like this:

`<s><s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi L<s>`

I also experienced something similar before that's why I made that connection. After changing that the training  process worked fine with the same text so I didn't look further into it and assumed that was the reason.

**Comment by FartyPants** [comment]
> @FartyPants I jumped to conclusions too fast for the reason of failing and my explanation is quite lackluster there. I definitely wasn't thorough enough explaining that part and shouldn't have mentioned that the train fails because of that.
> 
> Let me rephrase it, I assumed that the problem was with the BOS token, right now it will appear several times at the begging of the text and possibly again later as the chunking process will copy it if the flag is on. If I am not mistaken the training text will end looking up like this:
> 
> `<s><s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi L<s>`
> 
> I also experienced something similar before that's why I made that connection. After changing that the training process worked fine with the same text so I didn't look further into it and assumed that was the reason.

You may have point there, multiple encode/decode could also multiply the BOS token.
I'm looking at the dataset part of the prepare  # == Prep the dataset, format, etc ==

tokens = shared.tokenizer.encode(text_part) ->first encode
text_chunks = [shared.tokenizer.decode(x) for x in out_tokens] ->decode
train_data = Dataset.from_list([tokenize(x) for x in text_chunks]) ->Second encode

so in the dataset leg it seems we are adding two, but let me just run test so we are sure, be right back

The plaintext leg seems ok though

**Comment by FartyPants** [comment]
Let me do some series of test on that

**Comment by Nan-Do** [comment]
@FartyPants I just checked running the code and on my side the tokens are generated as I said previously. Perhaps you can try this to confirm it. I just added a couple of prints to the generation process.
``` python
    if raw_text_file not in ['None', '']:
        logger.info("Loading raw text file dataset...")

        train_template["template_type"] = "raw_text"

        with open(clean_path('training/datasets', f'{raw_text_file}.txt'), 'r', encoding='utf-8') as file:
            raw_text = file.read().replace('\r', '')

        cut_string = hard_cut_string.replace('\\n', '\n')
        out_tokens = []
        for text_part in raw_text.split(cut_string):
            if text_part.strip() == '':
                continue

            tokens = shared.tokenizer.encode(text_part)
            step = cutoff_len - overlap_len
            if step <= 0:
                yield f"Error: overlap_len ({overlap_len}) cannot be greater than or equal to cutoff_len ({cutoff_len})"
                return

            tokens = list(split_chunks(tokens, step))
            for i in range(1, len(tokens)):
                tokens[i] = tokens[i - 1][-overlap_len:] + tokens[i]

            out_tokens.extend(tokens)
            del tokens

        del raw_text  # Note: could be a gig for a large dataset, so delete redundant data as we go to be safe on RAM
        text_chunks = [shared.tokenizer.decode(x) for x in out_tokens]
        del out_tokens
        if newline_favor_len > 0:
            text_chunks = [cut_chunk_for_newline(x, newline_favor_len) for x in text_chunks]
        print(text_chunks)

        train_data = Dataset.from_list([tokenize(x) for x in text_chunks])
        print(train_data['input_ids'])
        print(list(map(shared.tokenizer.decode, train_data['input_ids'])))
        del text_chunks
        eval_data = None
```
Raw text:
```
Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostrud officia pariatur ut officia. Sit irure elit esse ea nulla sunt ex occaecat reprehenderit commodo officia dolor Lorem duis laboris cupidatat officia voluptate. Culpa proident adipisicing id nulla nisi laboris ex in Lorem sunt duis officia eiusmod. Aliqua reprehenderit commodo ex non excepteur duis sunt velit enim. Voluptate laboris sint cupidatat ullamco ut ea consectetur et est culpa et culpa duis.
```
Output:
```
['<s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostrud officia pariatur ut officia. Sit irure elit esse ea nulla sunt ex occaecat reprehenderit comm', '<s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostrud officia pariatur ut officia. Sit irure elit esse ea nulla sunt ex occaecat reprehenderit commodo officia dolor Lorem duis laboris cupidatat officia voluptate. Culpa proident adipisicing id nulla nisi laboris ex in Lorem sunt duis officia eiusmod. Aliqua reprehenderit commodo ex non excepteur duis sunt velit enim. Voluptate laboris sint cupidatat ullamco ut ea consectetur et est culpa et culpa duis.']
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 365, 3668, 23421, 22224, 7845, 27315, 29892, 3344, 423, 429, 13300, 371, 332, 429, 20924, 7163, 14767, 29882, 1581, 277, 427, 326, 9775, 487, 13949, 3274, 25507, 594, 302, 10770, 365, 3668, 610, 29875, 1337, 286, 3028, 277, 429, 831, 344, 11382, 7018, 27315, 29889, 405, 10770, 3778, 18002, 333, 271, 271, 429, 13300, 371, 332, 3344, 423, 29889, 830, 1457, 29882, 1581, 277, 20990, 29878, 566, 20990, 29878, 566, 23421, 365, 3668, 707, 394, 8105, 666, 27315, 1700, 21245, 403, 1700, 21245, 403, 22224, 6260, 302, 21797, 707, 410, 1693, 29889, 405, 520, 29878, 566, 3344, 423, 610, 29875, 1337, 3477, 3344, 423, 29889, 21988, 3805, 545, 560, 277, 831, 344, 321, 29874, 302, 21797, 23991, 429, 2179, 29874, 687, 271, 14767, 29882, 1581, 277, 844], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 365, 3668, 23421, 22224, 7845, 27315, 29892, 3344, 423, 429, 13300, 371, 332, 429, 20924, 7163, 14767, 29882, 1581, 277, 427, 326, 9775, 487, 13949, 3274, 25507, 594, 302, 10770, 365, 3668, 610, 29875, 1337, 286, 3028, 277, 429, 831, 344, 11382, 7018, 27315, 29889, 405, 10770, 3778, 18002, 333, 271, 271, 429, 13300, 371, 332, 3344, 423, 29889, 830, 1457, 29882, 1581, 277, 20990, 29878, 566, 20990, 29878, 566, 23421, 365, 3668, 707, 394, 8105, 666, 27315, 1700, 21245, 403, 1700, 21245, 403, 22224, 6260, 302, 21797, 707, 410, 1693, 29889, 405, 520, 29878, 566, 3344, 423, 610, 29875, 1337, 3477, 3344, 423, 29889, 21988, 3805, 545, 560, 277, 831, 344, 321, 29874, 302, 21797, 23991, 429, 2179, 29874, 687, 271, 14767, 29882, 1581, 277, 844, 8144, 3344, 423, 22224, 365, 3668, 868, 275, 10212, 275, 18002, 333, 271, 271, 3344, 423, 1700, 21245, 403, 29889, 315, 352, 3274, 410, 1693, 594, 666, 275, 18499, 1178, 302, 21797, 302, 10770, 10212, 275, 429, 297, 365, 3668, 23991, 868, 275, 3344, 423, 321, 2482, 1545, 29889, 10785, 21408, 14767, 29882, 1581, 277, 844, 8144, 429, 1661, 429, 13300, 371, 332, 868, 275, 23991, 5343, 277, 427, 326, 29889, 3684, 21245, 403, 10212, 275, 25507, 18002, 333, 271, 271, 318, 645, 314, 1111, 3477, 321, 29874, 5178, 312, 300, 332, 634, 707, 13949, 3274, 634, 13949, 3274, 868, 275, 29889]]
['<unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><s><s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostrud officia pariatur ut officia. Sit irure elit esse ea nulla sunt ex occaecat reprehenderit comm', '<unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><s><s>Lorem ipsum dolor sit amet, officia excepteur ex fugiat reprehenderit enim labore culpa sint ad nisi Lorem pariatur mollit ex esse exercitation amet. Nisi anim cupidatat excepteur officia. Reprehenderit nostrud nostrud ipsum Lorem est aliquip amet voluptate voluptate dolor minim nulla est proident. Nostrud officia pariatur ut officia. Sit irure elit esse ea nulla sunt ex occaecat reprehenderit commodo officia dolor Lorem duis laboris cupidatat officia voluptate. Culpa proident adipisicing id nulla nisi laboris ex in Lorem sunt duis officia eiusmod. Aliqua reprehenderit commodo ex non excepteur duis sunt velit enim. Voluptate laboris sint cupidatat ullamco ut ea consectetur et est culpa et culpa duis.']

**Comment by FartyPants** [comment]
Yes, I verified this in plaintext there is chance of adding `<s><s>`
The dataset leg seems ok.

`
 def tokenize(prompt):

        if train_only_after == '' or train_only_after not in prompt:
            input_ids = encode(prompt, True)
            input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
            labels = [1] * len(input_ids)

            eos_token_count = input_ids.count(shared.tokenizer.eos_token_id)
            if eos_token_count>0:
                print(f"EOS: {eos_token_count}")
            bos_token_count = input_ids.count(shared.tokenizer.bos_token_id)
            if bos_token_count>1:
                
                if input_ids[0]==input_ids[1]:
                    print(f"BOS: {bos_token_count} <s><s>")
                else:
                    print(f"BOS: {bos_token_count}")

`
EOS: 1
EOS: 1
BOS: 2 <s><s>
EOS: 1
EOS: 1
BOS: 2 <s><s>
EOS: 1
BOS: 2 <s><s>
EOS: 1
BOS: 2 <s><s>

**Comment by FartyPants** [comment]
> I'd like to invite @FartyPants for review
@oobabooga & @Nan-Do 

I verified, there is a chance of having block of text (plaintext) with  `<s><s>` due to multiple use of encode

I would suggest to add to def tokenize(prompt):

```
# Check if the first two tokens are BOS
if len(input_ids) >= 2 and input_ids[:2] == [shared.tokenizer.bos_token_id, shared.tokenizer.bos_token_id]:
    input_ids= input_ids[1:]
```

**Comment by Nan-Do** [comment]
Well, there are other changes in the PR that I'd like to see implemented, specially the change to LoRA.py right now models loaded in 4 or 8 bits can't use LoRAs. Of course, if that's the standard procedure I'll close the PR and send them separately

**Comment by FartyPants** [comment]
> Well, there are other changes in the PR that I'd like to see implemented, specially the change to LoRA.py right now models loaded in 4 or 8 bits can't use LoRAs. Of course, if that's the standard procedure I'll close the PR and send them separately

Ultimately it's @oobabooga who reviews these, so I would suggest to split it to two as those are two different things - and it is hard to review 2 things at once.

**Comment by Nan-Do** [comment]
I'll prepare the changes and send them separately when I have some free time. I'll try to make it by tomorrow, in the mean time I'll leave the PR open just in case @oobabooga  has something to say about it
